### PR TITLE
Fix bug in ShowerProcessor / HillasReconstructor not using QualityQuery correctly, add support for HillasIntersection in ShowerProcessor

### PR DESCRIPTION
**Pull request by maxnoe**
The code was requiring at least three! telescopes, while also the debug log output implied 2 and we should not hard code a minimum of 3 telescopes for stereo reconstruction.

### CODE REVIEW

**Review by LukasNickel — DISMISSED** [review]

**Review by nbiederbeck — DISMISSED** [review]

**Review by kosack — CHANGES_REQUESTED** [review]
This wasn't really meant to be a multiplicity cut for performance, but to just prevent events with less than 2 telescopes being processed (since the algorithm requires at least 2).    

Cuts on multiplicity should come later when doing the optimization or quality classification.   At the DL2 level, we should get all events, even ones with only 2 telescopes.    Later when making an IRF, you have the option to choose a tighter cut, however,

Remember we need to support event quality-types, so you could imagine putting all multiplicity-2 events in their own class. You don't want to drop them before that,

**Comment by LukasNickel** [comment]
The reason for changing this in the first place was that it applied a cut at 3 not 2 telescopes, despite the comments suggesting otherwise.
(> vs >=)

Event classes is a valid consideration though

**Comment by maxnoe** [comment]
@kosack I am not clear what you want us to change? You think the option should be removed and we should have a hard coded `>= 2`?

**Comment by kosack** [comment]
> @kosack I am not clear what you want us to change? You think the option should be removed and we should have a hard coded >= 2?

Yes, however I suppose we can leave it as an option, but the idea was not that this is a minimal cut so that the algorithm works, not the final cut for performance reasons.  Maybe just adding something to the help string would be enough?  I just want to avoid people thinking this is where they should apply the multiplicity cut, which in reality may be more complex and is applied in the DL2→DL3 step.

Fixing it to not require 3 is obviously  a good thing however!

**Review by kosack — APPROVED** [review]

**Comment by maxnoe** [comment]
> Yes, however I suppose we can leave it as an option, but the idea was *not* that this is a minimal cut so that the algorithm works, not the final cut for performance reasons. 

First not (emphasis mine) not intended?

**Comment by maxnoe** [comment]
I am fine with just checking `>= 2` and not making it configurable. Everything else can be done after.

**Comment by maxnoe** [comment]
After realizing the issue described here #1876, I went ahead and added the fix here.

With this, the `ShowerProcessor` is now able to be configured to use both `HillasIntersection` and `HillasReconstructor`, we do really check for `N < 2`, not `N > 2`, and the quality query is properly taken into account.

**Comment by maxnoe** [comment]
This is a breaking change now, since it changes the API of the reconstructors and the corresponding config files need to be adapted, anyways, let's make a quick new release and everybody should just skip 0.13

**Review by nbiederbeck — CHANGES_REQUESTED** [review]

File: ctapipe/reco/hillas_reconstructor.py [hunk 1]
```diff
@@ -211,13 +217,11 @@ def __call__(self, event):
...
                 event, hillas_dict, self.subarray, array_pointing, telescope_pointings
             )
-        except (TooFewTelescopesException, InvalidWidthException):
-            result = ReconstructedGeometryContainer()
-
-        event.dl2.stereo.geometry["HillasReconstructor"] = result
+        except InvalidWidthException:
+            return INVALID
 
     def _predict(
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply ctapipe/reco/hillas_reconstructor.py:225]: Only that in the refactoring done before, `__call__` was introduced only for `HillasReconstructor`, not for `HillasIntersection` and `predict` was made `_predict` here.

File: ctapipe/reco/hillas_intersection.py [hunk 2]
```diff
@@ -84,6 +88,43 @@ def __init__(self, config=None, parent=None, **kwargs):
...
             self._weight_method = self.weight_konrad
 
+        self.check_parameters = ShowerQualityQuery(parent=self)
+
+
+    def __call__(self, event):
+        hillas_dict = {
+            tel_id: dl1.parameters.hillas
+            for tel_id, dl1 in event.dl1.tel.items()
+            if all(self.check_parameters(dl1.parameters))
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply ctapipe/reco/hillas_intersection.py:98]: actually, `all` is faster for few elements

File: ctapipe/tools/tests/resources/stage2_config.json [hunk 3]
```diff
@@ -41,13 +41,15 @@
...
-                ["enough pixels", "lambda p: p.morphology.num_pixels > 3"],
-                ["not clipped", "lambda p: p.leakage.intensity_width_2 < 2"]
-            ]
+        "reconstructor_type": "HillasReconstructor",
+        "HillasReconstructor": {
+            "ShowerQualityQuery": {
+                "quality_criteria": [
+                    ["enough intensity", "lambda p: p.hillas.intensity > 50"],
+                    ["Positive width", "lambda p: p.hillas.width.value > 0"],
+                    ["enough pixels", "lambda p: p.morphology.num_pixels > 3"]
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply ctapipe/tools/tests/resources/stage2_config.json:50]: The check was `leakage < 2`, which is always true. I made all files consistent. We can talk about introducing a leakage cut by default, but I wouldn't do it here

[the remark this answers is not in the document]
Comment (nbiederbeck) [reply ctapipe/tools/tests/resources/stage2_config.json:50]: +1 for consistency
and sorry I didn't check the content of that line

File: ctapipe/reco/hillas_reconstructor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nbiederbeck) [reply ctapipe/reco/hillas_reconstructor.py:224]: Got it, thanks

File: ctapipe/reco/hillas_reconstructor.py [hunk 4]
```diff
@@ -256,17 +260,13 @@ class are set to np.nan
...
-            )
         # check for np.nan or 0 width's as these screw up weights
-        if any([np.isnan(hillas_dict[tel]["width"].value) for tel in hillas_dict]):
+        if any([np.isnan(h.width.value) for h in hillas_dict.values()]):
             raise InvalidWidthException(
                 "A HillasContainer contains an ellipse of width==np.nan"
             )
-        if any([hillas_dict[tel]["width"].value == 0 for tel in hillas_dict]):
+
+        if any([h.width.value == 0 for h in hillas_dict.values()]):
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply ctapipe/reco/hillas_reconstructor.py:262]: moved to `Reconstructor`

**Comment by maxnoe** [comment]
I think this is ready now and a big improvement. The code in both reconstructors could be a lot better though I think, but I'd keep that for another PR.

**Review by kosack — APPROVED** [review]
seems to work fine

**Review by nbiederbeck — APPROVED** [review]

File: ctapipe/reco/tests/test_hillas_intersection.py [hunk 5]
```diff
@@ -49,29 +47,29 @@ def test_parallel():
...
-    hill_inter = HillasIntersection()
+    hill_inter = HillasIntersection(subarray=example_subarray)
 
-    horizon_frame = AltAz()
+    altaz = AltAz()
     zen_pointing = 10 * u.deg
 
     array_direction = SkyCoord(
-        alt=90 * u.deg - zen_pointing, az=0 * u.deg, frame=horizon_frame
+        alt=90 * u.deg - zen_pointing, az=0 * u.deg, frame=altaz
```
[the remark this answers is not in the document]
Comment (kosack) [reply ctapipe/reco/tests/test_hillas_intersection.py:59]: You can also just write 'altaz' as a string I think.

**Comment by maxnoe** [comment]
I had to update the notebook, please review again. It's now also much more up-to-date which what is currently there in ctapipe (Using ImageProcessor, ShowerProcessor and DataWriter)

**Comment by maxnoe** [comment]
I processed 10 La Palma gamma-diffuse Prod5b runs, works very nicely. Comparison of current master vs. this branch:

multiplicity_shower_processor

theta2_shower_processor

**Comment by kosack** [comment]
Do you understand why there is a difference for multiplicity 3+ events? Shouldn't it be identical?

File: ctapipe/reco/hillas_reconstructor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (maxnoe) [reply ctapipe/reco/hillas_reconstructor.py]: @kosack width is divided by in the weighting, which is why the check is for exactly 0

**Comment by maxnoe** [comment]
> Do you understand why there is a difference for multiplicity 3+ events? Shouldn't it be identical?

I am checking now, I think I did not adapt the config correctly I used.

**Comment by maxnoe** [comment]
I indeed forgot to change `ShowerQualityQuery` to `StereoQualityQuery` in my config file. With that fixed, it matches my expectation:

**Review by nbiederbeck — APPROVED** [review]

**Pull request merged**